csharp/src/Google.Protobuf/IMessage.cs

```diff
@@ -69,6 +69,18 @@ public interface IMessage
         MessageDescriptor Descriptor { get; }
     }
 
+    /// <summary>
+    /// Interface for a Protocol Buffers message, supporting 
+    /// proto2 operations required for serialization
+    /// </summary>
+    public interface IMessage2 : IMessage

```

Comment:
What if we want to add other API for IMessage in future?

Reply:
If we really, *really* don't want to add new methods to the IMessage interface or make new interface versions, I could create a reflection based extension method that checks if a message is initialized that way.

Reply:
"making an abstract class and inheriting from that would break code that inherits from another class in a separate partial definition." I don't understand this clear.

@jtattermusch do you have any suggestions how to add a new API to IMessage interface without break compatibility tests?  I suggested to add an abstract in between IMessage and generated message, but seems like it has some issue. Add another interface in between can solve the problem but we still unable to extend IMessage APIs in future.

Reply:
To explain, the generated messages are partial. This allows for implementing OnConstruction in the generated code without modifying it. This also allows for users to make the generated code to extend upon a base class if they wish. Since classes in .NET can't inherit from two base classes, if you add an abstract class into the generated code this breaks any user's code that makes the generated code use another base class in a separate partial declaration.

For example:

Generated code file
```cs
public partial class Message : pb::GeneratedMessage
{
    // … generated code
}
```
User's code
```cs
public partial class Message : BaseClass // compiler error: Partial declarations of 'Message' must not specify different base classes
{
    // … some user code
}
```

Reply:
one important point: adding methods to an interface breaks backward compatibility (e.g. after adding a method to IMessage, the implementations of IMessage outside of this repository wont' compile anymore -> backward compatibility is broken), so adding a method to IMessage is not an option (unless we want to break the backward compatibility and release a new  major version, and I don't think we want that).

ad naming: "IMessage2" doesn't seem like a good name (it looks a bit like a hacked together offshoot of IMessage and it's not descriptive). Sth like "IMessageProto2" is at least descriptive.

Reply:
We should NOT break backward compatibility, thus can't add method into IMessage. Add a new interface is not a good solution, still unable to extend base methods in future. Add a new interface each time scares.   

"if you add an abstract class into the generated code this breaks any user's code that makes the generated code use another base class in a separate partial declaration." Other protobuf languages do not support such cases, from protobuf team's review, we can break such cases and document it at the release notes. However, we still want to know from C# devolopers' veiw. @jtattermusch and @jskeet, do you agree to break such cases?

Reply:
I would personally strongly prefer not to break users like this. Breaking users breaks trust, which can be very hard to regain later. It's possible that C# users have a lower tolerance to breaking than other developers, or it may be a personal bias.

It's not only breaking, but breaking in a way which is very hard to recover from. It's not like "You need to change method call X by adding another argument" or similar.

Default methods in C# 8 might help in the future, but it'll be a while before we can assume those...

Reply:
Agree to use field accessors for now before we can add the default methods